libc/runtime/zipos-open.c

```diff
@@ -79,13 +83,19 @@ static void *__zipos_mmap_space(size_t mapsize) {
 }
 
 struct ZiposHandle *__zipos_keep(struct ZiposHandle *h) {
-  atomic_fetch_add_explicit(&h->refs, 1, memory_order_relaxed);
+  int refs = atomic_fetch_add_explicit(&h->refs, 1, memory_order_relaxed);
+  if (VERY_UNLIKELY(refs > MAX_REFS)) {
+    abort();
+  }
   return h;
 }
 
 static bool __zipos_drop(struct ZiposHandle *h) {
-  int refs = atomic_load_explicit(&h->refs, memory_order_acquire);
-  return -1 == refs || -1 == atomic_fetch_sub(&h->refs, 1);
+  if (!atomic_fetch_sub_explicit(&h->refs, 1, memory_order_release)) {
```

Comment:
Interesting unref design. For posterity, I'd be interested in hearing where this is written.

Reply:
Sources cited in the commit message.